openedx/core/lib/xblock_keys/bundle_def.py

```diff
@@ -0,0 +1,139 @@
+        """
+        Convert an arbitrary OLX path to a potentially shorter format.
+        The result is shorter if the olx path is in the standard format
+        /block_type-block_id/block_id.olx
+
+        The result is a string like "block_type/block_id" or the original string.
+
+        Returns a tuple: short_string, type_part or None, id_part or None
+        """
+        if len(olx_path) < 2 or olx_path[0] != '/':
+            raise ValueError("An OLX path must start with / and contain a filename")
+        if olx_path.endswith('.olx'):
+            id_part = olx_path[:-4].rpartition('/')[2]  # from "/foo/bar.olx", get "bar"
+            type_part = olx_path[1:].partition('-' + id_part + '/')[0] # from "/foo-bar/bar.olx", get "foo"
+            short_path = type_part + '/' + id_part
+            if cls._olx_path_expanded(short_path)[0] == olx_path:
+                return (short_path, type_part, id_part)
+        return (olx_path, None, None)
+
+    @classmethod
+    def _olx_path_expanded(cls, olx_path_shortened):
+        """
+        Convert a potentially shortened OLX path to the full expanded form
+        The result is a string like "/block_type-block_id/block_id.olx"
+
+        Returns a tuple: olx_path, type_part or None, id_part or None
+        """
+        if len(olx_path_shortened) < 2:
+            raise ValueError("Invalid short OLX path")
+        if olx_path_shortened[0] == '/':
+            return (olx_path_shortened, None, None)  # It's a normal OLX path
+        try:
+            type_part, _, id_part = olx_path_shortened.partition('/')
+        except ValueError:
+            raise ValueError("Invalid short OLX path")
+        olx_path = '/' + type_part + '-' + id_part + '/' + id_part + '.olx'
+        return (olx_path, type_part, id_part)
+
+    @staticmethod
+    def _b64_encode(data):
+        """ Encode arbitrary bytes in a URL-friendly base64 format """
+        return base64.b64encode(data, '-_').rstrip('=')
+
+    @staticmethod
+    def _b64_decode(encoded):
+        """ Inverse of _b64_encode """
+        encoded = encoded + '=' * (3 - len(encoded) % 3)  # Fix padding
+        if isinstance(encoded, unicode):
+            encoded = encoded.encode('utf8')
+        return base64.b64decode(encoded, '-_')
+
+    def _to_string(self):
+        """
+        Return a string representing this location in blockstore.
+
+        The encoding is a bit complex in order to keep the string length down.
+        """
+        bundle_uuid_encoded = self._b64_encode(self.bundle_uuid.bytes)
+        olx_path_encoded, root_type, root_id = self._olx_path_shortened(self.olx_path)
+        block_type_encoded = '' if self.block_type == root_type else self.block_type
+        definition_id_encoded = '' if self.definition_id == root_id else self.definition_id
+
+        return u":".join((
+            bundle_uuid_encoded,
+            olx_path_encoded,
+            block_type_encoded,
+            definition_id_encoded,
+        ))
```

Comment:
I think it would be very nice if this string representation looked very similar to how OLX files/course outlines/pathways reference blocks in the bundles. :)

Reply:
I agree that would be nice, but it doesn't necessarily have to be the case.

**Definition IDs** are something used internally by the XBlock runtimes only, and there's no reason to think that they would ever appear in blockstore or be seen directly by users at all. (Usage keys on the other hand also wouldn't necessarily appear in blockstore but are sometimes seen by course authors.)

My thought was that since Blockstore needs to support many different kinds of references (e.g. references to `.html` files, references to PDF documents or SVG images, references to other bundles, references to other OLX files, _maybe_ but not necessarily references to specific blocks inside other OLX files), we would have one way of doing references (and links) within blockstore bundles, and another totally separate key mechanism for identifying XBlocks loaded by runtimes from blockstore.